### PR TITLE
Adds API retries for certain error codes

### PR DESCRIPTION
**Pull request by amcpanaligan**
# Notes
1. Wrapped `Q promises` into native `Promise`
2. Reworked `attempt` to accept `function` and invoke it from there than `Promise`

### CODE REVIEW

File: src/client/mixins/prospect-aware.ts [hunk 1]
```diff
@@ -1,41 +1,25 @@
...
+  public retry: any;
+
+  public attempt: (fn: (...params) => Promise<any>, retryCount: number, ...params) => Promise<any>;
 
   public async createProspect(prospect: Record<string, any>) {
     await this.clientReady;
-    return new Promise((resolve, reject) => {
-      this.client.prospects.create(prospect.email, prospect).then(resolve).fail(reject);
-    });
+    return this.attempt(this.client.prospects.create, 1, prospect.email, prospect);
```
Comment (iamEAP) [inline src/client/mixins/prospect-aware.ts]: Part of the difficulty I had reading this was never quite being sure if `this.client.prospects.create` was referring to the method on the `lew-pardot` client itself, or an instance of the `PardotClient` class.

```javascript
return this.attempt(this.client.prospects.create, 1, prospect.email, prospect);
```

We could eliminate the need for the intermediary wrapper class by putting the Q Promise wrapping logic in the attempt, using a simple anonymous function.

```javascript
return this.attempt(() => {
  return new Promise((resolve, reject) => {
    // Where this.client refers to an authenticated lew-pardot instance
    this.client.prospects.create(prospect.email, prospect).then(resolve).fail(reject);
  });
}, 1);
```

File: src/client/client-wrapper.ts [hunk 2]
```diff
@@ -19,25 +21,56 @@ class ClientWrapper {
...
+          reject('Login failed. Please check your auth credentials and try again.');
+        } else if (err.code === this.DAILY_API_LIMIT_EXCEEDED_ERROR_CODE) {
+          reject('API call limit reached for today.');
+        }
       });
     });
-
   }
 
+  public async attempt(fn: (...params) => Promise<any>, retryCount = 1, ...params) {
```
Comment (iamEAP) [inline src/client/client-wrapper.ts]: And if we used that anonymous function approach (inline below), we could also simplify the definition of the `attempt` method here.

```javascript
public async attempt(fn: () => Promise<any>, retryCount = 1) {
  // ...
  return new Promise ((resolve, reject) => {
    operation.attempt((currentAttempt: number) => {
      fn().then(resolve).catch((err: Error) => {
        // ...
      });
    });
  });
}
```

Comment (amcpanaligan) [reply src/client/client-wrapper.ts]: Does it work? By invoking only `fn()`? Will the parameter get passed in?

Comment (amcpanaligan) [reply src/client/client-wrapper.ts]: Ahhh! The outer code is changed as well to give a callback

File: src/client/mixins/prospect-aware.ts (hunk 1 above)
Comment (amcpanaligan) [reply src/client/mixins/prospect-aware.ts]: Ahh I just wrapped them as experiment of I can stub them that way.

**Pull request merged**